libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,336 @@
+
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            Subscription qSub1 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+            Subscription qSub2 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+
+            natsConnection.publish("mySubject", "data".getBytes());
+
+            // Simulate real world time for messages to propagate.
+            Awaitility.await()
+                .atMost(WAIT_AT_MOST_MILLIS, TimeUnit.MILLISECONDS)
+                .pollDelay(POLL_DELAY_MILLIS, TimeUnit.MILLISECONDS)
+                .until(() -> messages.size() == 1);
+
+            assertEquals(1, messages.size());
+            dispatcher.unsubscribe(qSub1);
+            dispatcher.unsubscribe(qSub2);
+        }
     }
 
+    @Test
+    public void whenMessagesAreExchangedViaPublish_thenResponsesMustBeReceivedWithSecondarySubscription() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            Subscription replySideSubscription1 = natsConnection.subscribe("publishSubject");
+            Subscription replySideSubscription2 = natsConnection.subscribe("publishSubject");
+
+            Subscription publishSideSubscription = natsConnection.subscribe("replyToSubject");
+            natsConnection.publish("publishSubject", "replyToSubject", "Please respond!".getBytes());
+
+            Message message = replySideSubscription1.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            assertEquals("Please respond!", new String(message.getData()));
+            natsConnection.publish(message.getReplyTo(), "Message Received By Subscription 1".getBytes());
+
+            message = replySideSubscription2.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            assertEquals("Please respond!", new String(message.getData()));
+            natsConnection.publish(message.getReplyTo(), "Message Received By Subscription 2".getBytes());
+
+            int responsesFrom1 = 0;
+            int responsesFrom2 = 0;
+            message = publishSideSubscription.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            String response = new String(message.getData());
+            assertTrue(response.contains("Message Received"));
+            if (response.contains("1")) {
+                responsesFrom1++;
+            }
+            else if (response.contains("2")) {
+                responsesFrom2++;
+            }
+
+            message = publishSideSubscription.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            response = new String(message.getData());
+            assertTrue(response.contains("Message Received"));
+            if (response.contains("1")) {
+                responsesFrom1++;
+            }
+            else if (response.contains("2")) {
+                responsesFrom2++;
+            }
```

Comment:
This can be written as a 2 element for loop like:

for (int i = 0; i < 2; i++ ) {
    // duplicated receive logic
}

Reply:
Why is that better? Do you want me to use an array instead of `responsesFrom1` and `responsesFrom2` I think this is clearer.

Reply:
It's unclear at a glance if this is the same pattern repeated twice, or there is some subtle difference, the loop idea (or a helper method) would clarify that it's the same

Reply:
fixed